File: tasks/b4-fastapi-greenfield-service/src/store.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Optional

from .schemas import TransactionCreate, TransactionResponse


@dataclass
class TransactionRecord:
    id: int
    amount: Decimal
    type: str
    description: Optional[str]


class TransactionStore:
    def __init__(self) -> None:
        self._records: list[TransactionRecord] = []
        self._next_id = 1

    def reset(self) -> None:
        self._records.clear()
        self._next_id = 1

    def create(self, payload: TransactionCreate) -> TransactionResponse:
        record = TransactionRecord(
            id=self._next_id,
            amount=payload.amount,
            type=payload.type,
            description=payload.description,
        )
        self._next_id += 1
        self._records.append(record)
        return TransactionResponse(
            id=record.id,
            amount=record.amount,
            type=record.type,  # type: ignore[arg-type]
            description=record.description,
        )

    def list_all(self) -> list[TransactionResponse]:
        return [
            TransactionResponse(
                id=record.id,
                amount=record.amount,
                type=record.type,  # type: ignore[arg-type]
                description=record.description,
            )
            for record in self._records
        ]

    def balance(self) -> Decimal:
        total = Decimal("0")
        for record in self._records:
            if record.type == "credit":
                total += record.amount
            else:
                total -= record.amount
        return total


store = TransactionStore()

File: tasks/b5-nodejs-greenfield-api-or-cli/src/store.ts
import { formatMoney } from "./money.js";
import type { TransactionCreate, TransactionResponse } from "./schemas.js";

interface TransactionRecord {
  id: number;
  amount: string;
  type: "credit" | "debit";
  description?: string;
}

class TransactionStore {
  private records: TransactionRecord[] = [];
  private nextId = 1;

  reset(): void {
    this.records = [];
    this.nextId = 1;
  }

  create(payload: TransactionCreate): TransactionResponse {
    const record: TransactionRecord = {
      id: this.nextId,
      amount: payload.amount,
      type: payload.type,
      description: payload.description,
    };
    this.nextId += 1;
    this.records.push(record);
    return { ...record };
  }

  listAll(): TransactionResponse[] {
    return this.records.map((record) => ({ ...record }));
  }

  balance(): string {
    const total = this.records.reduce((sum, record) => {
      const amount = Number(record.amount);
      return record.type === "credit" ? sum + amount : sum - amount;
    }, 0);
    return formatMoney(total);
  }
}

export const store = new TransactionStore();

export function resetStore(): void {
  store.reset();
}
